Consolidate flushes even when no read in progress

olim7t opened the pull request:
Motivation:

Currently FlushConsolidationHandler only consolidates if a read loop is
active for a Channel, otherwise each writeAndFlush(...) call will still
be flushed individually. When these calls are close enough, it can be
beneficial to consolidate them even outside of a read loop.

Modifications:

When we allow a flush to "go through", don't perform it immediately, but
schedule it on the channel's executor with a timeout of 0. Under high
pressure, this gives other writes a chance to enqueue before the task
gets executed, and so we flush multiple writes at once.

Result:

Lower CPU usage and less context switching.

olim7t commented on handler/src/main/java/io/netty/handler/flush/FlushConsolidationHandler.java:
```diff
@@ -70,17 +75,21 @@ public FlushConsolidationHandler(int explicitFlushAfterFlushes) {
 
     @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
```
If my understanding is correct, this is right as long as the handler is not shared.

normanmaurer replied:
Yes

normanmaurer commented:
Can you share a benchmark or some numbers?

normanmaurer commented on handler/src/main/java/io/netty/handler/flush/FlushConsolidationHandler.java:
```diff
@@ -132,14 +141,32 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+
+    private void scheduleFlush(final ChannelHandlerContext ctx) {
+        // Run as soon as possible, but still yield to give a chance for additional writes to enqueue.
+        nextScheduledFlush = ctx.channel().eventLoop().schedule(flushTask, 0, TimeUnit.MILLISECONDS);
```
I think using executor().submit(...) is a better choice

normanmaurer replied:
@olim7t shouldn't this be:

```java
if (nextScheduleFlush == null) {
    nextScheduledFlush = ctx.channel().eventLoop().schedule(flushTask, 0, TimeUnit.MILLISECONDS);
}
```

Otherwise you will mist to cancel all scheduled flushes once `flushIfNeeded` is called. Beside this I think there is no need to schedule multiple if the one scheduled did not run yet.

olim7t commented:
@normanmaurer Let me check with my management what I can share publicly.

normanmaurer commented:
Thx if not you know how to find me

> Am 09.09.2016 um 19:19 schrieb Olivier Michallat <notifications@github.com>:
> 
> @normanmaurer Let me check with my management what I can share publicly.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

normanmaurer commented on handler/src/main/java/io/netty/handler/flush/FlushConsolidationHandler.java:
```diff
@@ -132,14 +141,32 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        nextScheduledFlush = ctx.channel().eventLoop().schedule(flushTask, 0, TimeUnit.MILLISECONDS);
+    }
+
+    private Runnable flushTask = new Runnable() {
```
final ?

normanmaurer replied:
also can you move this to the beginning of the class please :)

normanmaurer commented on handler/src/main/java/io/netty/handler/flush/FlushConsolidationHandler.java:
```diff
@@ -132,14 +141,32 @@ public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+            if (flushPendingCount > 0 && !readInProgress) {
+                flushPendingCount = 0;
+                ctx.flush();
+            } // else we'll flush when the read completes
```
Also `nextScheduledFlush = null;`

normanmaurer commented on handler/src/test/java/io/netty/handler/flush/FlushConsolidationHandlerTest.java:
```diff
@@ -27,12 +27,40 @@
 
 public class FlushConsolidationHandlerTest {
 
+    static final int EXPLICIT_FLUSH_AFTER_FLUSHES = 3;
```
private

normanmaurer commented on handler/src/test/java/io/netty/handler/flush/FlushConsolidationHandlerTest.java:
```diff
@@ -27,12 +27,40 @@
+        assertEquals(0, flushCount.get());
+        // Trigger the execution of the async task
+        channel.runPendingTasks();
+        assertEquals(1, flushCount.get());
```
Also call `channel.finish()` and assert the return value.

normanmaurer commented on handler/src/test/java/io/netty/handler/flush/FlushConsolidationHandlerTest.java:
```diff
@@ -27,12 +27,40 @@
+        for (int i = 0; i < EXPLICIT_FLUSH_AFTER_FLUSHES; i++) {
+            channel.flush();
+        }
+        assertEquals(1, flushCount.get());
```
Also call `channel.finish()` and assert the return value.

olim7t commented:
I addressed all your comments. Sorry for the delay, I wanted to run a few more tests.

I thought I saw a difference between `submit` and `schedule`, but it was just noise in my tests, over a few runs the differences are not significant.

I do however see a clear difference in favor of the [custom flusher](https://github.com/datastax/java-driver/blob/3.0/driver-core/src/main/java/com/datastax/driver/core/Connection.java#L849) we are currently using in the Datastax Java driver: `FlushConsolidationHandler` triggers approximately 15% more context switches, and uses a bit more CPU. One thing that the Java driver code does differently is that it coalesces not only the `flush()` call, but also the writes themselves: nothing is written to the channel until we decide to flush (the messages are enqueued). Before we flush, we execute all the pending writes from the channel's event loop. I think doing the writes sequentially instead of submitting them as individual tasks could explain the reduced context switching, does that make sense?

olim7t commented:
Actually I realize the same behavior could be achieved in the handler by overriding `write`, I'll give it a try.

netkins commented on handler/src/main/java/io/netty/handler/flush/FlushConsolidationHandler.java:
```diff
@@ -23,20 +23,23 @@
 import io.netty.channel.ChannelPromise;
 
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'java.util.concurrent.TimeUnit'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

olim7t commented:
> Actually I realize the same behavior could be achieved in the handler by overriding write, I'll give it a try.

Forget that last comment, processing a write in the handler implies that it's been submitted as a task already, so that wouldn't solve the issue.

Scottmitch commented:
good idea @olim7t ! lgtm

normanmaurer commented:
@olim7t can you please squash your commits ?

normanmaurer commented:
@olim7t will merge soon. Just want to make one more test.

olim7t commented:
Sounds good to me, I will try to work on that today.

normanmaurer commented on handler/src/main/java/io/netty/handler/flush/FlushConsolidationHandler.java:
```diff
@@ -46,41 +56,81 @@
+    private ChannelHandlerContext ctx;
+    private Future<?> nextScheduledFlush;
+
+    private final Runnable flushTask = new Runnable() {
```
use `null` if `consolidateWhenNoReadInProgress` is false ?

Scottmitch commented on handler/src/main/java/io/netty/handler/flush/FlushConsolidationHandler.java:
```diff
@@ -46,41 +56,83 @@
-                ctx.flush();
+                flushNow(ctx);
+            }
+        } else {
```
promote the inner `if` to make this an `else if` for less branch statements.

Scottmitch commented:
@normanmaurer - providing this as optional behavior sgtm.

normanmaurer commented:
@olim7t awesome! Squash and I will merge it :)

olim7t commented:
Squashed and rebased on current 4.0.

normanmaurer commented:
Cherry-picked into 4.0 (11f09bb5a6bac48c1da73cba3442652f8a8e6a1d) and 4.1 (3a8b8c9219901bb3fe6dd8b8b87b41eec3e275fc)

normanmaurer commented:
@olim7t thanks!
